core/src/main/java/org/elasticsearch/common/geo/builders/PolygonBuilder.java

```diff
@@ -482,7 +482,7 @@ private static void assign(Edge[] holes, Coordinate[][] points, int numHoles, Ed
                             && !(sharedVertex = (edges[pos].intersect.compareTo(current.coordinate) == 0)) ) {
                 throw new InvalidShapeException("Invalid shape: Hole is not within polygon");
             }
-            final int index = -((sharedVertex) ? 0 : pos+2);
+            final int index = sharedVertex || pos == -1 ? 0 : -(pos+2);
```

Comment:
@DaveCTurner thanks for walking me trough this f2f, it took some time for me to understand but now I think I understand this whole method a lot better. This looks right to me now after some longer thought, I think we should add some comments to this part after that explains the three(?) cases that might be true at this point and what that means for the index of the edge we assign this component to.